File: src/Template/XmlSequenceWrapperTemplate.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Net;
using System.Text;
using AutoRest.Extensions;
using AutoRest.Core.Utilities;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using AutoRest.Java.Model;
using System.Text.RegularExpressions;

namespace AutoRest.Java
{
    /// <summary>
    /// Writes an XmlSequenceWrapper to a JavaFile.
    /// </summary>
    public class XmlSequenceWrapperTemplate : IJavaTemplate<XmlSequenceWrapper, JavaFile>
    {
        private static XmlSequenceWrapperTemplate _instance = new XmlSequenceWrapperTemplate();
        public static XmlSequenceWrapperTemplate Instance => _instance;

        private XmlSequenceWrapperTemplate()
        {
        }

        public void Write(XmlSequenceWrapper xmlSequenceWrapper, JavaFile javaFile)
        {
            string xmlRootElementName = xmlSequenceWrapper.XmlRootElementName;
            string xmlListElementName = xmlSequenceWrapper.XmlListElementName;

            string xmlElementNameCamelCase = xmlRootElementName.ToCamelCase();

            ListType sequenceType = xmlSequenceWrapper.SequenceType;

            javaFile.Import(xmlSequenceWrapper.Imports);

            javaFile.JavadocComment(comment =>
            {
                comment.Description($"A wrapper around {sequenceType} which provides top-level metadata for serialization.");
            });
            javaFile.Annotation($"JacksonXmlRootElement(localName = \"{xmlRootElementName}\")");
            javaFile.PublicFinalClass(xmlSequenceWrapper.WrapperClassName, classBlock =>
            {
                classBlock.Annotation($"JacksonXmlProperty(localName = \"{xmlListElementName}\")");
                classBlock.PrivateFinalMemberVariable(sequenceType.ToString(), xmlElementNameCamelCase);

                classBlock.JavadocComment(comment =>
                {
                    comment.Description($"Creates an instance of {xmlSequenceWrapper.WrapperClassName}.");
                    comment.Param(xmlElementNameCamelCase, "the list");
                });
                classBlock.Annotation("JsonCreator");
                classBlock.PublicConstructor($"{xmlSequenceWrapper.WrapperClassName}(@JsonProperty(\"{xmlListElementName}\") {sequenceType} {xmlElementNameCamelCase})", constructor =>
                {
                    constructor.Line($"this.{xmlElementNameCamelCase} = {xmlElementNameCamelCase};");
                });

                classBlock.JavadocComment(comment =>
                {
                    comment.Description($"Get the {sequenceType} contained in this wrapper.");
                    comment.Return($"the {sequenceType}");
                });
                classBlock.PublicMethod($"{sequenceType} items()", function =>
                {
                    function.Return(xmlElementNameCamelCase);
                });
            });
        }
    }
}

File: mapper/src/main/java/com/azure/autorest/template/XmlSequenceWrapperTemplate.java
package com.azure.autorest.template;


// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.


import com.azure.autorest.model.clientmodel.ListType;
import com.azure.autorest.model.clientmodel.XmlSequenceWrapper;
import com.azure.autorest.model.javamodel.JavaFile;
import com.azure.autorest.util.CodeNamer;

/**
 Writes an XmlSequenceWrapper to a JavaFile.
*/
public class XmlSequenceWrapperTemplate implements IJavaTemplate<XmlSequenceWrapper, JavaFile>
{
    private static XmlSequenceWrapperTemplate _instance = new XmlSequenceWrapperTemplate();
    public static XmlSequenceWrapperTemplate getInstance()
    {
        return _instance;
    }

    private XmlSequenceWrapperTemplate()
    {
    }

    public final void Write(XmlSequenceWrapper xmlSequenceWrapper, JavaFile javaFile)
    {
        String xmlRootElementName = xmlSequenceWrapper.getXmlRootElementName();
        String xmlListElementName = xmlSequenceWrapper.getXmlListElementName();

        String xmlElementNameCamelCase = CodeNamer.toCamelCase(xmlRootElementName);

        ListType sequenceType = xmlSequenceWrapper.getSequenceType();

        javaFile.Import(xmlSequenceWrapper.getImports());

        javaFile.JavadocComment(comment ->
        {
                comment.Description(String.format("A wrapper around %1$s which provides top-level metadata for serialization.", sequenceType));
        });
        javaFile.Annotation(String.format("JacksonXmlRootElement(localName = \"%1$s\")", xmlRootElementName));
        javaFile.PublicFinalClass(xmlSequenceWrapper.getWrapperClassName(), classBlock ->
        {
                classBlock.Annotation(String.format("JacksonXmlProperty(localName = \"%1$s\")", xmlListElementName));
                classBlock.PrivateFinalMemberVariable(sequenceType.toString(), xmlElementNameCamelCase);

                classBlock.JavadocComment(comment ->
                {
                    comment.Description(String.format("Creates an instance of %1$s.", xmlSequenceWrapper.getWrapperClassName()));
                    comment.Param(xmlElementNameCamelCase, "the list");
                });
                classBlock.Annotation("JsonCreator");
                classBlock.PublicConstructor(String.format("%1$s(@JsonProperty(\"%2$s\") %3$s %4$s)", xmlSequenceWrapper.getWrapperClassName(), xmlListElementName, sequenceType, xmlElementNameCamelCase), constructor ->
                {
                    constructor.Line(String.format("this.%1$s = %2$s;", xmlElementNameCamelCase, xmlElementNameCamelCase));
                });

                classBlock.JavadocComment(comment ->
                {
                    comment.Description(String.format("Get the %1$s contained in this wrapper.", sequenceType));
                    comment.Return(String.format("the %1$s", sequenceType));
                });
                classBlock.PublicMethod(String.format("%1$s items()", sequenceType), function ->
                {
                    function.Return(xmlElementNameCamelCase);
                });
        });
    }
}